lib/extended-css-parser.js

```diff
@@ -0,0 +1,206 @@
+     */
+    var addQuotes = function (_, c1, c2) { 
+        return ':' + c1 + '("' + c2.replace(/["\\]/g, '\\$&') + '")';
+    };
+
+    /**
+     * Normalizes specified css text in a form that can be parsed by the
+     * Sizzle engine.
+     * Normalization means 
+     *  1. transforming [-ext-*=""] attributes to pseudo classes
+     *  2. enclosing possibly ambiguous arguments of `:contains`,
+     *     `:matches-css` pseudo classes with quotes.
+     * @param {string} cssText
+     * @return {string}
+     */
+    var normalize = function (cssText) {
+        cssText = cssText.replace(reAttrFallback, evaluateMatch);
+        cssText = cssText.replace(reMatchesCss, addQuotes);
+        cssText = cssText.replace(reContains, addQuotes);
+        return cssText;
+    };
+
+    /**
+     * Keeps the offset of `cssText` with respect to its original value,
+     * to provide a meaningful error position.
+     */
+    var posOffset;    
+    /**
+     * Throws an error.
+     */
+    var parseError = function (position) {
+        throw new Error('CssParser: parse error at position ' + (posOffset + position));
+    };
+
+    var reDeclEnd = /[;}]/g;
+    var reDeclDivider = /[;:}]/g;
+    var reNonWhitespace = /\S/g;
+
+    /**
+     * Parses css text into ExtendedSelector instances and style declarations.
+     * @param {string} cssText
+     * @return {Array<{selector:ExtendedSelector, style:Object<string,string>>}
+     */
+    var parseCss = function (cssText) {
+        if (!cssText) {
+            parseError(0);
+        }
+        cssText = normalize(cssText);
+        var cssLength = cssText.length;
+        var results = []; // An array to be returned.
+
+        posOffset = 0;
+
+        // Start of the stylesheet parse loop
+        while (cssText) {
+            // Apply tolerant tokenization.
+            var parseResult = Sizzle.tokenize(cssText, false, true);
+            if (typeof parseResult === 'string') {
+                parseError(cssLength);
+            }
+            var selector = parseResult.selector;
+            var tokens = parseResult.tokens;
+            var nextIndex = parseResult.nextIndex;
+
+            if (cssText.charCodeAt(nextIndex) !== 123 /* charCode of '{' */) {
+                parseError(nextIndex);
+            }
+
+            // Selector may end with whitespaces, and Sizzle
+            // treat it as a descendant combinator.
+            // We strip it out.
+            var lastTokens = tokens[tokens.length - 1];
+            var l = lastTokens.length;
+
+            if (lastTokens[l - 1].type === ' ') {
+                lastTokens.length = l - 1;
+            }
+            selector = selector.trim();
+
+            // Start of style declaration parse loop 
+            var styles = Object.create(null);
+            var match, matchPos, matched;
+            nextIndex++; // Move the pointer to the start of style declaration.
+            while (true) {
```

Comment:
Avoid `while (true)` loops at any cost. There's always a chance of an infinite loop.

If you can't find a better way, use `for (i < SOME_HUGE_NUMBER)` and throw an error if you didn't get the result in the end.